### DISCUSSION TITLE
How to return more information when a rule was triggered

### DISCUSSION THREAD

**Discussion by Edo3Bytes** [asulwer/RulesEngine#88] [Q&A]
We are planning on using v6.0.5 (or higher) of the RuleEngine but we'd like to return 2 values for each Rule that was successful: a Score and a Reason. We want to add all the Scores together after all Rules are processed, and display a Reason for each rule that was successful.  Of course we can return something like "5|rule 25 was triggered" and "10|rule 23 was triggered" for SuccessEvent and parse the Score out of that but that feels hacky. 

Any tips?

**Comment by tomwolanski** [comment]
One possible workaround I found is to use a custom action to create a result record on successful rule execution.
The problem is, that actions to not have knowledge about the rule, so the name of the rule has to be passed as an extra parameter to the action. This drastically increases the chances of the introduction of bugs and becomes hard to maintain.

Consider a workflow:
```json
{
  "WorkflowName": "Discount",
  "Rules": [
    {
	"RuleName": "MyRule",
	"Expression": "input1.country == \"india\"  AND input1.totalPurchasesToDate >= 5000 ",
      	"Actions": {
    		"OnSuccess": {
	       		"Name": "MyCustomAction", 
       			"Context": { 
				"ruleName" : "MyRule",              // <----- extra place where a rule name has to be added
				"otherValue": "this is result",
				"complexValue": {
					"s": "sValue",
					"i": 7
				}
       			}
    		}
 	}
  ]
}
```

The action can use any of the values added to the context and any of the parameters and initialize any type you'd like to return. The final result is available by `RuleResultTree.ActionResult`

```csharp
public class MyCustomAction : ActionBase
{
	public override async ValueTask<object> Run(ActionContext context, RuleParameter[] ruleParameters)
	{
		var ruleName = context.GetContext<string>("ruleName");
		var otherValue = context.GetContext<string>("otherValue");
		var complex = context.GetContext<ComplexType>("complexValue");

		return new { ruleName, otherValue, complex };
	}
}

async Task Main()
{
	var reSettings = new ReSettings
	{
		CustomActions = new Dictionary<string, Func<ActionBase>>()
		{
			["MyCustomAction"] = () => new MyCustomAction(),
		}
	};

	var re = new RulesEngine.RulesEngine(workflowRules, reSettings);
	var resultList = await re.ExecuteAllRulesAsync("Discount", new SomeInput("india", 50001));
		
	resultList.Dump();
}

```

Full snippet: https://dotnetfiddle.net/SOM6XB
